### PR TITLE
feat(jupyter): support Deno.test()

### PR DESCRIPTION
**Pull request by nayeemrmn**
Blocked by #20783.

image

image

### CODE REVIEW

**Comment by rgbkrk** [comment]
I am so excited about this because now we can give tutorials on writing tests within a notebook. 👨🏻‍🍳💋

File: cli/tools/repl/session.rs [hunk 1]
```diff
@@ -337,7 +353,58 @@ impl ReplSession {
       self.evaluate_ts_expression(line).await
     } else {
       evaluate_response
+    };
+
+    let mut test_reporter = PrettyTestReporter::new(false, true, false, true);
+    {
+      use crate::tools::jupyter::server::StdioMsg;
+      use tokio::sync::mpsc::UnboundedSender;
```
[the remark this answers is not in the document]
Comment (nayeemrmn) [reply cli/tools/repl/session.rs]: I extracted it with `ReplSession::set_test_reporter_factory()`. It's a bit crude but simple enough for a stopgap short of properly unifying output capturing between repl, testing and jupyter.

File: cli/tools/jupyter/mod.rs [hunk 2]
```diff
@@ -71,13 +80,25 @@ pub async fn kernel(
...
+  let stderr = StdioPipe::File(test_event_sender.stderr());
   let mut worker = worker_factory
     .create_custom_worker(
       main_module.clone(),
       permissions,
-      vec![ops::jupyter::deno_jupyter::init_ops(stdio_tx)],
-      Default::default(),
+      vec![
+        ops::jupyter::deno_jupyter::init_ops(stdio_tx.clone()),
+        crate::ops::testing::deno_test::init_ops(test_event_sender.clone()),
```
Comment (bartlomieju) [inline cli/tools/jupyter/mod.rs:94]:
```suggestion
        ops::testing::deno_test::init_ops(test_event_sender.clone()),
```

File: cli/tools/jupyter/mod.rs [hunk 3]
```diff
@@ -71,13 +80,25 @@ pub async fn kernel(
...
+      ],
+      // FIXME(nayeemrmn): Test output capturing currently doesn't work.
```
Comment (bartlomieju) [inline cli/tools/jupyter/mod.rs:96]: Hm, shouldn't we be able to capture the test output the same way `deno test` does it? Isn't it a matter of forwarding it to a proper channel?

**Review by bartlomieju — COMMENTED** [review]
Overall looks good, nearly there. I have one question regarding output capture

File: cli/tools/jupyter/mod.rs (hunk 3 above)
Comment (nayeemrmn) [reply cli/tools/jupyter/mod.rs:96]: It causes all logged output in the repl to be buffered and then flushed to stdout at once when a test event is processed.

**Review by bartlomieju — APPROVED** [review]
LGTM, let's try to address the capturing soon

**Pull request merged**